### ISSUE TITLE
QLoRA perfomance degradation after loading with PeftModel.from_pretrained

### ISSUE THREAD

**Issue by vanerk03** [huggingface/peft#1815]
### System Info

`peft==0.11.1`
`accelerate==0.30.1`
`transformers==4.41.2`
`python3.10`

### Reproduction

I noticed that the score degrades drastically after loading from checkpoint compared to training (the score on validation set the trainer reports) (for a classification task logloss went from 0.96 to 0.98, which is significant, but still not exceptionally bad, which means it's likely due to the rounding error). 

My assumption is that it might be because, by default, from_pretrained loads some of the weights (classifier head, for instance) in float16, although it is kept as float32 during training and is stored in float32 in `adapter_model.safetensors`.

I could not figure out a way to load everything directly in float32 and therefore did not check check if this is indeed due to the rounding error.

I am using the following code for training:
```
import os
import pandas as pd
import json
from pprint import pprint
import torch

import bitsandbytes as bnb
import torch.nn as nn
import transformers
from datasets import load_dataset, Dataset
from huggingface_hub import notebook_login

from peft import LoraConfig, PeftConfig, PeftModel, get_peft_model, prepare_model_for_kbit_training
from transformers import AutoConfig, AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig
from transformers import AutoModelForSequenceClassification

model = "meta-llama/Meta-Llama-3-8B"

MODEL_NAME = model

bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_use_double_quant=False,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.bfloat16
)

model = AutoModelForSequenceClassification.from_pretrained(
    MODEL_NAME,
    device_map="auto",
    trust_remote_code=True,
    quantization_config=bnb_config,
    num_labels=3,
    torch_dtype=torch.float16
)

tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)
tokenizer.pad_token = tokenizer.eos_token

config = LoraConfig(
    r=16,
    lora_alpha=32,
    target_modules=["q_proj", "k_proj", "v_proj", "o_proj", "down_proj", "gate_proj", "up_proj"],
    lora_dropout=0.05,
    bias="none",
    task_type="SEQ_CLS",
)

model = prepare_model_for_kbit_training(model)
model = get_peft_model(model, config)


# Training
training_args = transformers.TrainingArguments(
    per_device_train_batch_size=4,
    gradient_accumulation_steps=3,

    num_train_epochs=1,
    learning_rate=5e-5,
    bf16=True,
    
    eval_strategy="steps",
    eval_steps=500,
    logging_steps=500,
    save_steps=500,

    greater_is_better=False,
    # load_best_model_at_end=True,
    save_total_limit=5,
    metric_for_best_model='eval_loss',
    

    output_dir="llama_ft",
    optim="paged_adamw_8bit",
    
    lr_scheduler_type="linear",
    warmup_ratio=0.06,
    report_to="none",
)

train_tokenized, test_tokenized = ...

trainer = transformers.Trainer(
    model=model,
    train_dataset=train_tokenized,
    eval_dataset=test_tokenized,
    args=training_args,
    tokenizer=tokenizer
)

trainer.train()
```

The scipt I use for loading the checkpoint:
```
model = "meta-llama/Meta-Llama-3-8B"

MODEL_NAME = model

bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_use_double_quant=False,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.bfloat16
)

model = AutoModelForSequenceClassification.from_pretrained(
    MODEL_NAME,
    device_map="auto",
    trust_remote_code=True,
    quantization_config=bnb_config,
    num_labels=3,
    torch_dtype=torch.float16
)

tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)
tokenizer.pad_token = tokenizer.eos_token

model = PeftModel.from_pretrained(
    model,
    "./llama_ft/checkpoint-3500",
)

training_args = transformers.TrainingArguments(
    # same args as in the previous scipt
)

train_tokenized, test_tokenized = ... # I double checked that I use the same datasets 

trainer = transformers.Trainer(
    model=model,
    train_dataset=train_tokenized,
    eval_dataset=test_tokenized,
    args=training_args,
    tokenizer=tokenizer
)

trainer.evaluate() # now evaluation
``` 

Here is how `adapter_config.json` looks like:
```
{
  "alpha_pattern": {},
  "auto_mapping": null,
  "base_model_name_or_path": "meta-llama/Meta-Llama-3-8B",
  "bias": "none",
  "fan_in_fan_out": false,
  "inference_mode": true,
  "init_lora_weights": true,
  "layer_replication": null,
  "layers_pattern": null,
  "layers_to_transform": null,
  "loftq_config": {},
  "lora_alpha": 32,
  "lora_dropout": 0.05,
  "megatron_config": null,
  "megatron_core": "megatron.core",
  "modules_to_save": [
    "classifier",
    "score"
  ],
  "peft_type": "LORA",
  "r": 16,
  "rank_pattern": {},
  "revision": null,
  "target_modules": [
    "v_proj",
    "up_proj",
    "o_proj",
    "gate_proj",
    "down_proj",
    "q_proj",
    "k_proj"
  ],
  "task_type": "SEQ_CLS",
  "use_dora": false,
  "use_rslora": false
}
```

### Expected behavior

Score during finetuning should not differ from loading the checkpoint. (again, I double-checked that I am using the same test set).

**Comment by BenjaminBossan** [comment]
Your dtypes probably differ because for training, you're calling `prepare_model_for_kbit_training`. Could you check if you get the same dtypes if you call this method consistently (i.e. either for both or not at all)?

If this is the thing that makes the difference, you could cast the dtypes yourself, after all this is what the function does under the hood.

**Comment by vanerk03** [comment]
I tried to call `prepare_model_for_kbit_training` right before `PeftModel.from_pretrained`, in that case dtypes do match (all of the floating parameters are now in fp32), but the quality still degrades.

**Comment by BenjaminBossan** [comment]
Okay, that should not happen. Normally, it should be possible to recreate the exact same outputs after reloading. Here is a demonstration based on your code:

```python
import torch

from peft import LoraConfig, PeftConfig, PeftModel, get_peft_model, prepare_model_for_kbit_training
from transformers import AutoConfig, AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig
from transformers import AutoModelForSequenceClassification

MODEL_NAME = "facebook/opt-125m"

bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_use_double_quant=False,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.bfloat16
)

model = AutoModelForSequenceClassification.from_pretrained(
    MODEL_NAME,
    device_map="auto",
    trust_remote_code=True,
    quantization_config=bnb_config,
    num_labels=3,
    torch_dtype=torch.float16
)

config = LoraConfig(
    r=16,
    lora_alpha=32,
    target_modules=["q_proj", "k_proj", "v_proj", "o_proj", "down_proj", "gate_proj", "up_proj"],
    lora_dropout=0.05,
    bias="none",
    task_type="SEQ_CLS",
    init_lora_weights=False,  # <= to simulate a trained LoRA adapter
)

model = prepare_model_for_kbit_training(model)
model = get_peft_model(model, config)
model.eval()
input = torch.tensor([[1, 2, 3, 4, 5]])
output = model(input).logits
print("outputs before")
print(output)

model.save_pretrained("/tmp/peft/1815")
del model

model = AutoModelForSequenceClassification.from_pretrained(
    MODEL_NAME,
    device_map="auto",
    trust_remote_code=True,
    quantization_config=bnb_config,
    num_labels=3,
    torch_dtype=torch.float16
)
model = prepare_model_for_kbit_training(model)
model = PeftModel.from_pretrained(model, "/tmp/peft/1815")

output = model(input).logits
print("outputs after")
print(output)
torch.testing.assert_close(output, output)  # does not raise
```

Also, could you please show the repr of the model?

**Comment by vanerk03** [comment]
I apologise, I made a typo in my evaluation script that resulted in performance drop. Thank you for fast response, sorry again for the disturbance.
